fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java

```diff
@@ -858,42 +886,103 @@ protected void addTenantKey() {
+                                + "AND EXISTS (SELECT 1 FROM fhir_admin.tenant_keys tk WHERE tk.mt_id = t.mt_id "
+                                + "AND tk.tenant_hash = sysibm.hash(tk.tenant_salt || ?, 2));";
+                try (PreparedStatement stmt = connectionPool.getConnection().prepareStatement(sql)) {
+                    stmt.setString(1, tenantName);
+                    stmt.setString(2, tenantKey);
+                    if (stmt.execute()) {
+                        try (ResultSet resultSet = stmt.getResultSet();) {
+                            if (resultSet.next()) {
+                                throw new IllegalArgumentException("tenantName and tenantKey already exists");
+                            }
+                        }
+                    } else {
+                        throw new IllegalArgumentException("Problem checking the results");
+                    }
+                } catch (SQLException e) {
+                    throw new IllegalArgumentException(
+                            "Exception when querying backend to verify tenant key and tenant name", e);
+                }
+            } catch (DataAccessException x) {
+                // Something went wrong, so mark the transaction as failed
+                tx.setRollbackOnly();
+                throw x;
+            }
+        }
+    }
+
     /**
      * Allocate this tenant, creating new partitions if required.
      */
     protected void allocateTenant() {
-        if (dbType == DbType.DERBY || dbType == DbType.POSTGRESQL) {
+        if (!MULTITENANT_FEATURE_ENABLED.contains(dbType)) {
             return;
         }
+
         // The key we'll use for this tenant. This key should be used in subsequent
         // activities related to this tenant, such as setting the tenant context.
-        final String tenantKey = getRandomKey();
+        if (tenantKeyFileUtil.keyFileExists(tenantKeyFileName)) {
+            // Only if the Tenant Key file is provided as a parameter is it not null. 
+            // in  this case we want special behavior.
+            tenantKey = this.tenantKeyFileUtil.readTenantFile(tenantKeyFileName);
+        } else {
+            tenantKey = getRandomKey();
+        }
 
         // The salt is used when we hash the tenantKey. We're just using SHA-256 for
         // the hash here, not multiple rounds of a password hashing algorithm. It's
         // sufficient in our case because we are using a 32-byte random value as the
         // key, giving 256 bits of entropy.
         final String tenantSalt = getRandomKey();
 
+        Db2Adapter adapter = new Db2Adapter(connectionPool);
+        checkIfTenantNameAndTenantKeyExists(adapter, tenantName, tenantKey);
+
         // Open a new transaction and associate it with our connection pool. Remember
         // that we don't support distributed transactions, so all connections within
         // this transaction must come from the same pool
```

Comment:
maybe move these below the if-else block so its closer to the try-with-resource?

Reply:
resolving conversation so I can edit in another suggestion